eth-contracts/contracts/service/ServiceProviderFactory.sol

```diff
@@ -272,4 +312,60 @@ contract ServiceProviderFactory is RegistryContract {
             _endpoint
         );
     }
+
```

Comment:
@hareeshnagaraj i wrote some comments for each function to just summarize each function as i was going through it. i added them below but feel free to ignore/modify